File: plugins/plugin-browser/python/elizaos_browser/actions/type.py
import logging
from typing import Any

from elizaos_browser.services.browser_service import BrowserService
from elizaos_browser.types import ActionResult
from elizaos_browser.utils.errors import (
    ActionError,
    SessionError,
    handle_browser_error,
)
from elizaos_browser.utils.url import parse_type_action

logger = logging.getLogger(__name__)


TYPE_ACTION = {
    "name": "BROWSER_TYPE",
    "similes": ["TYPE_TEXT", "INPUT", "ENTER_TEXT"],
    "description": "Type text into an input field on the webpage",
    "examples": [
        {
            "user": 'Type "hello world" in the search box',
            "agent": 'I\'ve typed "hello world" in the search box.',
        }
    ],
}


async def browser_type(
    service: BrowserService,
    message: str,
    callback: Any | None = None,
) -> ActionResult:
    session = await service.get_or_create_session()
    if not session:
        error = SessionError("No active browser session")
        handle_browser_error(error, callback, "type text")
        return ActionResult(
            success=False,
            error="no_session",
            data={"actionName": "BROWSER_TYPE"},
        )

    text_to_type, field = parse_type_action(message)

    if not text_to_type:
        raise ActionError("type", field, ValueError("No text specified to type"))

    result = await service.get_client().type_text(session.id, text_to_type, field)
    if not result.success:
        raise ActionError("type", field, RuntimeError(result.error or "Type failed"))

    response_text = f'I\'ve typed "{text_to_type}" in the {field}'
    if callback:
        callback({"text": response_text, "actions": ["BROWSER_TYPE"]})

    return ActionResult(
        success=True,
        data={
            "actionName": "BROWSER_TYPE",
            "textTyped": text_to_type,
            "field": field,
            "sessionId": session.id,
        },
    )

File: plugins/plugin-browser/typescript/src/actions/type.ts
import type {
  Action,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { ServiceType } from "@elizaos/core";
import type { BrowserService } from "../services/browser-service.js";
import {
  ActionError,
  handleBrowserError,
  ServiceNotAvailableError,
  SessionError,
} from "../utils/errors.js";

export const browserTypeAction: Action = {
  name: "BROWSER_TYPE",
  similes: ["TYPE_TEXT", "INPUT", "ENTER_TEXT"],
  description: "Type text into an input field on the webpage",

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const browserEnabled =
      runtime.getSetting("ENABLE_BROWSER") === "true" ||
      runtime.getSetting("BROWSER_ENABLED") === "true";

    if (!browserEnabled) {
      return false;
    }

    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      return false;
    }

    const text = message.content?.text?.toLowerCase() ?? "";
    return text.includes("type") || text.includes("input") || text.includes("enter");
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult | undefined> => {
    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      const error = new ServiceNotAvailableError();
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "type text"
      );
      return {
        text: "Browser service is not available",
        success: false,
        data: {
          actionName: "BROWSER_TYPE",
          error: "service_not_available",
        },
        values: {
          success: false,
          errorType: "service_not_available",
        },
      };
    }

    const session = await service.getOrCreateSession();
    if (!session) {
      const error = new SessionError("No active browser session");
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "type text"
      );
      return {
        text: "No active browser session",
        success: false,
        data: {
          actionName: "BROWSER_TYPE",
          error: "no_session",
        },
        values: {
          success: false,
          errorType: "no_session",
        },
      };
    }

    const text = message.content?.text ?? "";
    const match = text.match(/["']([^"']+)["']/);
    const textToType = match ? match[1] : "";

    const fieldMatch = text.match(/(?:in|into) (?:the )?(.+)$/i);
    const field = fieldMatch ? fieldMatch[1] : "input field";

    if (!textToType) {
      throw new ActionError("type", field, new Error("No text specified to type"));
    }

    const result = await service.getClient().type(session.id, textToType, field);
    if (!result.success) {
      throw new ActionError("type", field, new Error(result.error ?? "Type failed"));
    }

    const responseContent: Content = {
      text: `I've typed "${textToType}" in the ${field}`,
      actions: ["BROWSER_TYPE"],
      source: message.content?.source ?? "action",
    };

    await callback?.(responseContent);

    return {
      text: responseContent.text ?? "",
      success: true,
      data: {
        actionName: "BROWSER_TYPE",
        textTyped: textToType,
        field,
        sessionId: session.id,
      },
      values: {
        success: true,
        textTyped: textToType,
        field,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user}}",
        content: { text: 'Type "hello world" in the search box' },
      },
      {
        name: "{{agent}}",
        content: {
          text: 'I\'ve typed "hello world" in the search box.',
          actions: ["BROWSER_TYPE"],
        },
      },
    ],
  ],
};
